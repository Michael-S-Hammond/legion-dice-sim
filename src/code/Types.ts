export enum DieColor {
    Red = "red",
    Black = "black",
    White = "white",
}

export enum AttackDieResult {
    Critical = 1,
    Hit = 2,
    Surge = 3,
    Miss = 4,
}

export type AttackRoll = {
  color: DieColor,
  result: AttackDieResult,
}

export enum AttackSurgeConversion {
  Blank = 1,
  Hit = 2,
  Critical = 3,
}

export enum Cover {
  None = "none",
  Light = "light",
  Heavy = "heavy",
}

export enum DefenseDieResult {
  Block = 1,
  Surge = 2,
  Blank = 3,
}

export type AbilityX = {
  active: boolean,
  value: number,
}

export type OffenseTokens = {
  aim: number,
  surge: number,
}

export type OffenseInput = {
  redDice: number,
  blackDice: number,
  whiteDice: number,
  surge: AttackSurgeConversion,
  tokens: OffenseTokens,
  blast: boolean,
  criticalX: AbilityX,
  duelist: boolean,
  highVelocity: boolean,
  impactX: AbilityX,
  ionX: AbilityX,
  jediHunter: boolean,
  lethalX: AbilityX,
  makashiMastery: boolean,
  pierceX: AbilityX,
  preciseX: AbilityX,
  ramX: AbilityX,
  sharpshooterX: AbilityX,
}

export type DefenseTokens = {
  dodge: number,
  observation: number,
  shield: number,
  suppression: number
  surge: number,
}

export type DefenseInput = {
  dieColor: DieColor,
  surge: boolean,
  tokens: DefenseTokens,
  cover: Cover,
  armor: boolean,
  armorX: AbilityX,
  block: boolean,
  dangerSenseX: AbilityX,
  deflect: boolean,
  djemSoMastery: boolean,
  duelist: boolean,
  hasForceUpgrades: boolean,
  immuneBlast: boolean,
  immuneMeleePierce: boolean,
  immunePierce: boolean,
  impervious: boolean,
  lowProfile: boolean,
  outmaneuver: boolean,
  soresuMastery: boolean,
  uncannyLuckX: AbilityX,
}

export type CombatInput = {
  guardian: AbilityX,
  meleeAttack: boolean,
};

export type AttackInput = {
  offense: OffenseInput,
  defense: DefenseInput,
  combat: CombatInput,
}

export type AttackResult = {
  criticals: number,
  hits: number,
  surges: number,
  misses: number,
}

export type DefenseResult = {
  forcedSaves: number,
  blocks: number,
  surges: number,
  blanks: number,
  wounds: number,
}

export type AttackOutput = {
  attack: AttackResult,
  defense: DefenseResult,
}

export type Stats = {
  mean: number,
  median: number,
  stddev: number,
}

export type AttackSummary = {
  attackCount: number,
  critical: Array<number>,
  hit: Array<number>,
  attackSurge: Array<number>,
  forcedSaves: Array<number>,
  forcedSaveStats: Stats,
  blocks: Array<number>,
  defenseSurge: Array<number>,
  wounds: Array<number>,
  woundStats: Stats
}

export type CombinedAttackOutput = {
  firstAttack: AttackOutput,
  summary: AttackSummary,
}

export enum ResultOutput {
  None = 1,
  Single = 2,
  Graph = 3,
}

export function createDefaultAttackInput() : AttackInput {
  return {
      offense: {
          redDice: 0,
          blackDice: 0,
          whiteDice: 0,
          surge: AttackSurgeConversion.Blank,
          tokens: {
            aim: 0,
            surge: 0,
          },
          blast: false,
          criticalX: { active: false, value: 1 },
          duelist: false,
          highVelocity: false,
          impactX: { active: false, value: 1 },
          ionX: { active: false, value: 1 },
          jediHunter: false,
          lethalX: { active: false, value: 1 },
          makashiMastery: false,
          pierceX: { active: false, value: 1 },
          preciseX: { active: false, value: 1 },
          ramX: { active: false, value: 1 },
          sharpshooterX: { active: false, value: 1 },
      },
      defense: {
          dieColor: DieColor.White,
          surge: true,
          tokens: {
            dodge: 0,
            observation: 0,
            shield: 0,
            suppression: 0,
            surge: 0,
          },
          cover: Cover.None,
          armor: false,
          armorX: { active: false, value: 1 },
          block: false,
          dangerSenseX: { active: false, value: 1 },
          deflect: false,
          djemSoMastery: false,
          duelist: false,
          hasForceUpgrades: false,
          immuneBlast: false,
          immuneMeleePierce: false,
          immunePierce: false,
          impervious: false,
          lowProfile: false,
          outmaneuver: false,
          soresuMastery: false,
          uncannyLuckX: { active: false, value: 1 }
      },
      combat: {
        guardian: { active: false, value: 1 },
        meleeAttack: false,
      },
  }
}

export function cloneAttackInput(original: AttackInput) : AttackInput {
  return {
      offense: {
        redDice: original.offense.redDice,
        blackDice: original.offense.blackDice,
        whiteDice: original.offense.whiteDice,
        surge: original.offense.surge,
        tokens: {
          aim: original.offense.tokens.aim,
          surge: original.offense.tokens.surge,
        },
        blast: original.offense.blast,
        criticalX: {
          active: original.offense.criticalX.active,
          value: original.offense.criticalX.value
        },
        duelist: original.offense.duelist,
        highVelocity: original.offense.highVelocity,
        impactX: {
          active: original.offense.impactX.active,
          value: original.offense.impactX.value
        },
        ionX: {
          active: original.offense.ionX.active,
          value: original.offense.ionX.value
        },
        jediHunter: original.offense.jediHunter,
        lethalX: {
          active: original.offense.lethalX.active,
          value: original.offense.lethalX.value
        },
        makashiMastery: original.offense.makashiMastery,
        pierceX: {
          active: original.offense.pierceX.active,
          value: original.offense.pierceX.value
        },
        preciseX: {
          active: original.offense.preciseX.active,
          value: original.offense.preciseX.value
        },
        ramX: {
          active: original.offense.ramX.active,
          value: original.offense.ramX.value,
        },
        sharpshooterX: {
          active: original.offense.sharpshooterX.active,
          value: original.offense.sharpshooterX.value
        },
    },
    defense: {
        dieColor: original.defense.dieColor,
        surge: original.defense.surge,
        tokens: {
          dodge: original.defense.tokens.dodge,
          observation: original.defense.tokens.observation,
          shield: original.defense.tokens.shield,
          suppression: original.defense.tokens.suppression,
          surge: original.defense.tokens.surge,
        },
        cover: original.defense.cover,
        armor: original.defense.armor,
        armorX: {
          active: original.defense.armorX.active,
          value: original.defense.armorX.value
        },
        block: original.defense.block,
        dangerSenseX: {
          active: original.defense.dangerSenseX.active,
          value: original.defense.dangerSenseX.value
        },
        deflect: original.defense.deflect,
        djemSoMastery: original.defense.djemSoMastery,
        duelist: original.defense.duelist,
        hasForceUpgrades: original.defense.hasForceUpgrades,
        immuneBlast: original.defense.immuneBlast,
        immuneMeleePierce: original.defense.immuneMeleePierce,
        immunePierce: original.defense.immunePierce,
        impervious: original.defense.impervious,
        lowProfile: original.defense.lowProfile,
        outmaneuver: original.defense.outmaneuver,
        soresuMastery: original.defense.soresuMastery,
        uncannyLuckX: {
          active: original.defense.uncannyLuckX.active,
          value: original.defense.uncannyLuckX.value
        },
    },
    combat: {
      guardian: {
        active: original.combat.guardian.active,
        value: original.combat.guardian.value
      },
      meleeAttack: original.combat.meleeAttack,
    },
  };
}
