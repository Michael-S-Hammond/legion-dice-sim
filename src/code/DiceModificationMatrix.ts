import * as T from "./Types";

const iCrit = 0;
const iHit = 1;
const iSurge = 2;
const iMiss = 3;

export type DieConversion = {
    roll: T.AttackRoll,
    conversion: T.AttackDieResult,
}

type MatchCriteriaCount = {
    roll: T.AttackRoll,
    count: number,
};

class DiceModificationMatrix {
    private _rolls: T.AttackRoll[];
    private matrices = {
        red:    [[0, 0, 0, 0],
                 [0, 0, 0, 0],
                 [0, 0, 0, 0],
                 [0, 0, 0, 0]],
        black:  [[0, 0, 0, 0],
                 [0, 0, 0, 0],
                 [0, 0, 0, 0],
                 [0, 0, 0, 0]],
        white:  [[0, 0, 0, 0],
                 [0, 0, 0, 0],
                 [0, 0, 0, 0],
                 [0, 0, 0, 0]],
    };

    constructor(rolls: T.AttackRoll[]) {
        this._rolls = rolls;
        for(let i = 0; i < rolls.length; i++) {
            const matrix = this.mtx4c(rolls[i].color);
            const index = this.idx4ar(rolls[i].result);
            matrix[index][index]++;
        }
    }

    public getRerollIndexes(hits: number, misses: number) : number[] {
        const indexes: number[] = [];
        const criteria : MatchCriteriaCount[] = [];
        let remainingMiss = misses;
        let remainingHits = hits;

        // TODO: improve this to take natural misses of the same color over converted misses
        const matchList = this.getConversions().filter((element: DieConversion, idx: number) : boolean => {
            if(element.conversion === T.AttackDieResult.Miss && remainingMiss > 0) {
                remainingMiss--;
                return true;
            }
            if(element.conversion === T.AttackDieResult.Hit && remainingHits > 0) {
                remainingHits--;
                return true;
            }
            return false;
        });

        matchList.forEach((element: DieConversion) => {
            let found = false;
            for(let i = 0; i < criteria.length; i++) {
                if(criteria[i].roll.color === element.roll.color &&
                    criteria[i].roll.result === element.roll.result) {
                    found = true;
                    criteria[i].count++;
                    break;
                }
            }

            if(!found) {
                criteria.push({
                    roll: {
                        color: element.roll.color,
                        result: element.roll.result,
                    },
                    count: 1,
                })
            }
        });

        this._rolls.forEach((element: T.AttackRoll, idx: number) => {
            for(let i = 0; i < criteria.length; i++) {
                if(criteria[i].roll.color === element.color &&
                    criteria[i].roll.result === element.result) {
                    if(criteria[i].count > 0) {
                        indexes.push(idx);
                        criteria[i].count--;
                    }
                    break;
                }
            }
        });

        return indexes;
    }

    public getResultCount(result: T.AttackDieResult) : number {
        const index = this.idx4ar(result);

        // count = original + sum(column) - sum(row)
        let count = this.matrices.red[index][index] +
                    this.matrices.black[index][index] +
                    this.matrices.white[index][index];
        for(let i = 0; i < 4; i++) {
            // subtract the values of the row (dice that have been converted from expected)
            count -= this.matrices.red[index][i] +
                     this.matrices.black[index][i] +
                     this.matrices.white[index][i];
            // add the values of the column (dice that have been converted to expected)
            count += this.matrices.red[i][index] +
                     this.matrices.black[i][index] +
                     this.matrices.white[i][index];
        }

        return count;
    }

    public getConversions() : DieConversion[] {
        const result : DieConversion[] = [];

        this.addConversions(T.DieColor.Red, result);
        this.addConversions(T.DieColor.Black, result);
        this.addConversions(T.DieColor.White, result);

        return result;
    }

    private addConversions(color: T.DieColor, conversions: DieConversion[]) {
        const matrix = this.mtx4c(color);
        for(let row = 0; row < 4; row++) {
            for(let col = 0; col < 4; col++) {
                let count = 0;
                if(row === col) {
                    count = (2 * matrix[row][col]) - this.sumRow(row, matrix);
                } else {
                    count = matrix[row][col];
                }

                for(let i = 0; i < count; i++) {
                    conversions.push({
                        roll: {
                            color: color,
                            result: this.idx2atk(row),
                        },
                        conversion: this.idx2atk(col),
                    })
                }
            }
        }
    }

    public tryConvertResult(originalResult: T.AttackDieResult, newResult: T.AttackDieResult) : boolean {
        let converted = false;
        if(this.isConvertingUp(originalResult, newResult)) {
            converted =
                this.tryConvertResultForColor(T.DieColor.White, originalResult, newResult) ||
                this.tryConvertResultForColor(T.DieColor.Black, originalResult, newResult) ||
                this.tryConvertResultForColor(T.DieColor.Red, originalResult, newResult);
        } else {
            converted =
                this.tryConvertResultForColor(T.DieColor.Red, originalResult, newResult) ||
                this.tryConvertResultForColor(T.DieColor.Black, originalResult, newResult) ||
                this.tryConvertResultForColor(T.DieColor.White, originalResult, newResult);
        }
        return converted;
    }

    public tryConvertResultCount(oldResult: T.AttackDieResult, newResult: T.AttackDieResult, count: number) : number {
        let converted = 0;
        for(let i = 0; i < count; i++) {
            if(!this.tryConvertResult(oldResult, newResult)) {
                break;
            }
            converted++;
        }
        return converted;
    }

    public tryConvertResultAllExcept(oldResult: T.AttackDieResult, newResult: T.AttackDieResult, maintain: number) : number {
        let converted = 0;
        let remaining = this.getResultCount(oldResult);
        while(remaining > maintain) {
            if(!this.tryConvertResult(oldResult, newResult)) {
                break;
            }
            converted++;
            remaining = this.getResultCount(oldResult);
        }
        return converted;
    }

    private getResultCountInMatrix(matrix: number[][], result: T.AttackDieResult) : number {
        const index = this.idx4ar(result);
        return matrix[index][index] + this.sumColumn(index, matrix) - this.sumRow(index, matrix);
    }

    private tryConvertResultForColor(color: T.DieColor, originalResult: T.AttackDieResult, newResult: T.AttackDieResult) : boolean {
        const matrix = this.mtx4c(color);
        const oridx = this.idx4ar(originalResult);
        const nridx = this.idx4ar(newResult);

        let converted = false;
        if(matrix[nridx][oridx] > 0) {
            matrix[nridx][oridx]--;
            converted = true;
        } else if((2 * matrix[oridx][oridx]) - this.sumRow(oridx, matrix) > 0) {
            matrix[oridx][nridx]++;
            converted = true;
        } else if(this.getResultCountInMatrix(matrix, originalResult) > 0) {
            for(let i = 3; i >= 0; i--) {
                if(i === oridx || i === nridx) {
                    // we checked these in the previous cases
                    continue;
                }
                if(matrix[i][oridx] > 0) {
                    matrix[i][oridx]--;
                    matrix[i][nridx]++;
                    converted = true;
                    break;
                }
            }
        }
        return converted;
    }

    private mtx4c(color: T.DieColor) : number[][] {
        if(color === T.DieColor.Red) {
            return this.matrices.red;
        }
        if(color === T.DieColor.Black) {
            return this.matrices.black;
        }
        return this.matrices.white;
    }

    private idx4ar(result: T.AttackDieResult) : number {
        return (result === T.AttackDieResult.Critical) ? iCrit :
                (result === T.AttackDieResult.Hit) ? iHit :
                (result === T.AttackDieResult.Surge) ? iSurge : iMiss;
    }

    private idx2atk(index: number) : T.AttackDieResult {
        return (index === iCrit) ? T.AttackDieResult.Critical :
                (index === iHit) ? T.AttackDieResult.Hit :
                (index === iSurge) ? T.AttackDieResult.Surge :
                T.AttackDieResult.Miss;
    }

    private sumRow(row: number, matrix: number[][]) {
        let sum = 0;
        for(let i = 0; i < 4; i++) {
            sum += matrix[row][i];
        }
        return sum;
    }

    private sumColumn(column: number, matrix: number[][]) {
        let sum = 0;
        for(let i = 0; i < 4; i++) {
            sum += matrix[i][column];
        }
        return sum;
    }

    private isConvertingUp(originalResult: T.AttackDieResult, newResult: T.AttackDieResult) : boolean {
        return newResult === T.AttackDieResult.Critical ||
                originalResult === T.AttackDieResult.Miss ||
                (originalResult === T.AttackDieResult.Surge && newResult !== T.AttackDieResult.Miss);
    }
}

export default DiceModificationMatrix;
