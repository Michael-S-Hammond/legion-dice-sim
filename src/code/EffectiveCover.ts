import * as T from './Types';

export function getCoverModification(cover: T.Cover) : number {
    return (cover === T.Cover.Heavy) ? 2 :
        (cover === T.Cover.Light) ? 1 :
        0;
}

export function getEffectiveCover(input: T.AttackInput) : T.Cover {
    // cover only applies to ranged attacks
    if(input.combat.meleeAttack) {
        return T.Cover.None;
    }

    // effective cover with modifications
    let effectiveCover = input.defense.cover;

    // cover from surpression
    if(input.defense.tokens.suppression > 0) {
        if(effectiveCover == T.Cover.None) {
            effectiveCover = T.Cover.Light;
        } else if(effectiveCover == T.Cover.Light) {
            effectiveCover = T.Cover.Heavy;
        }
    }

    // low profile
    if(input.defense.lowProfile && effectiveCover === T.Cover.Light) {
        effectiveCover = T.Cover.Heavy;
    }

    // sharpshooter X
    if(input.offense.sharpshooterX.active && effectiveCover !== T.Cover.None) {
        if(input.offense.sharpshooterX.value == 1) {
            if(effectiveCover === T.Cover.Heavy) {
                effectiveCover = T.Cover.Light;
            } else {
                effectiveCover = T.Cover.None;
            }
        } else if(input.offense.sharpshooterX.value == 2) {
            effectiveCover = T.Cover.None;
        }
    }

    // blast & immune: blast
    if(input.offense.blast && !input.defense.immuneBlast) {
        effectiveCover = T.Cover.None;
    }

    return effectiveCover;
}
