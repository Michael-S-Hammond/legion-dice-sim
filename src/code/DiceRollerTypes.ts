import * as T from './Types';
import DiceModificationMatrix from './DiceModificationMatrix';

export type AttackResultStatus = {
    attackResult: T.AttackResult,
    modificationMatrix: DiceModificationMatrix,
    defenseResult: T.DefenseResult,
    modifiedDefenseResult: ModifiedDefenseResult,
};

type ModifiedDefenseResult = {
    blocks: number,
    surges: number,
    blanks: number,
};
