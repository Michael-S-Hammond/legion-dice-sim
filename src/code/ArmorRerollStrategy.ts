import * as T from "./Types";
import * as RS from "./RerollStrategy";
import * as RSH from "./RerollStrategyHelpers";

import DiceModificationMatrix, * as DMM from "./DiceModificationMatrix";
import * as EC from './EffectiveCover';

class ArmorRerollStrategy implements RS.RerollStrategy {
    pickRerollDice(input: T.AttackInput,
                   rolls: T.AttackRoll[],
                   rerollDiceCount: number,
                   reason: RS.RerollReason,
                   remaining: RS.RemainingRerolls) : number[] | undefined {
        const dmm = new DiceModificationMatrix(rolls);
        const effectiveCover = EC.getEffectiveCover(input);
        const coverModifier = EC.getCoverModification(effectiveCover);
        let rerollMiss = 0;

        // convert surges
        if(input.offense.surge === T.AttackSurgeConversion.Critical) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, rolls.length);
        }
        if(input.offense.criticalX.active) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, input.offense.criticalX.value);
        }
        if(input.offense.surge === T.AttackSurgeConversion.Hit) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Hit, rolls.length);
        }
        dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Miss, rolls.length);

        // ram
        if(input.offense.ramX.active && input.combat.meleeAttack) {
            let ramRemaining = input.offense.ramX.value;
            ramRemaining -= dmm.tryConvertResultCount(T.AttackDieResult.Miss, T.AttackDieResult.Critical, ramRemaining);
            dmm.tryConvertResultCount(T.AttackDieResult.Hit, T.AttackDieResult.Critical, ramRemaining);
        }

        // any hits to be rerolled should be converted to misses in this block
        // therefore, hits should always be 0 when this completes

        const totalHits = dmm.getResultCount(T.AttackDieResult.Hit);
        let totalCrits = dmm.getResultCount(T.AttackDieResult.Critical);
        const possibleHits = Math.min(rolls.length - totalCrits, totalHits + RSH.getMaximumRerolledDice(input, remaining));
        let hitsToKeep = 0;
        if(input.offense.impactX.active && (possibleHits > coverModifier)) {
            hitsToKeep = Math.min(totalHits, coverModifier + input.offense.impactX.value);
        }

        // remove hits for cover and armor, less impact
        dmm.tryConvertResultAllExcept(T.AttackDieResult.Hit, T.AttackDieResult.Miss, hitsToKeep);

        // calculate misses that can be rerolled
        rerollMiss = Math.min(rerollDiceCount, dmm.getResultCount(T.AttackDieResult.Miss));

        if(reason === RS.RerollReason.AimToken) {
            // convert hits to crit for impact
            const impactConvert = Math.max(0, dmm.getResultCount(T.AttackDieResult.Hit) - coverModifier);
            dmm.tryConvertResultCount(T.AttackDieResult.Hit, T.AttackDieResult.Critical, impactConvert);

            // lethal
            totalCrits = dmm.getResultCount(T.AttackDieResult.Critical);
            const effectiveLethal = RSH.getEffectiveLethal(input, 0, totalCrits, 0);
            if(effectiveLethal > 0 && remaining.aimTokens <= effectiveLethal) {
                rerollMiss = 0;
            }
        }

        // return the result
        const indexes = dmm.getRerollIndexes(0, rerollMiss);
        return indexes.length == 0 ? undefined : indexes;
    }

    shouldHandle(input: T.AttackInput, reason: RS.RerollReason) : boolean {
        return input.defense.armor;
    }
}

export default ArmorRerollStrategy;
