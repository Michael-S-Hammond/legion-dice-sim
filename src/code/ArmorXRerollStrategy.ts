import * as T from "./Types";
import * as RS from "./RerollStrategy";
import * as RSH from "./RerollStrategyHelpers";

import DiceModificationMatrix from "./DiceModificationMatrix";
import * as EC from './EffectiveCover';

class ArmorXRerollStrategy implements RS.RerollStrategy {
    pickRerollDice(input: T.AttackInput,
                   rolls: T.AttackRoll[],
                   rerollDiceCount: number,
                   reason: RS.RerollReason,
                   remaining: RS.RemainingRerolls) : number[] | undefined {
        const dmm = new DiceModificationMatrix(rolls);
        const effectiveCover = EC.getEffectiveCover(input);
        const coverModifier = EC.getCoverModification(effectiveCover);
        let rerollMiss = 0;
        let rerollHit = 0;

        // convert surges
        if(input.offense.surge === T.AttackSurgeConversion.Critical) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, rolls.length);
        }
        if(input.offense.criticalX.active) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, input.offense.criticalX.value);
        }
        if(input.offense.surge === T.AttackSurgeConversion.Hit) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Hit, rolls.length);
        }
        dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Miss, rolls.length);

        rerollMiss = Math.min(rerollDiceCount, dmm.getResultCount(T.AttackDieResult.Miss));
        const startingHits = dmm.getResultCount(T.AttackDieResult.Hit);
        const totalCrits = dmm.getResultCount(T.AttackDieResult.Critical);
        const possibleHits = Math.min(rolls.length - totalCrits, startingHits + RSH.getMaximumRerolledDice(input, remaining));
        const hitsCancelledByCover = Math.min(startingHits, coverModifier);
        const hitsAfterCover = startingHits - hitsCancelledByCover;

        // convert for impact
        let effectiveImpact = 0;
        let remainingImpact = 0;
        let hitsAfterImpact = hitsAfterCover;
        if(input.offense.impactX.active) {
            effectiveImpact = Math.min(hitsAfterCover, input.offense.impactX.value);
            remainingImpact = input.offense.impactX.value - effectiveImpact;
            hitsAfterImpact = hitsAfterCover - effectiveImpact;
        }

        // convert for ram x
        let hitsAfterRam = hitsAfterImpact;
        let ramConvertMiss = 0;
        let ramConvertHit = 0;
        if(input.offense.ramX.active && input.combat.meleeAttack) {
            let remainingRam = input.offense.ramX.value;
            ramConvertMiss = Math.min(remainingRam, rerollMiss);
            rerollMiss -= ramConvertMiss;
            remainingRam -= ramConvertMiss;
            ramConvertHit = Math.min(remainingRam, hitsAfterImpact);
            hitsAfterRam -= ramConvertHit;
        }
        const effectiveRam = ramConvertMiss + ramConvertHit;

        // convert for armor x
        const remainingArmor = Math.max(0, input.defense.armorX.value - hitsAfterRam);
        const effectiveArmor = input.defense.armorX.value - remainingArmor;
        const hitsAfterArmor = hitsAfterRam - effectiveArmor;

        // determine effect of lethal
        let lethalOverride = false;
        if(reason === RS.RerollReason.AimToken) {
            const effectiveLethal = RSH.getEffectiveLethal(input, hitsAfterArmor, totalCrits + effectiveImpact, 0);
            if(effectiveLethal > 0 && remaining.aimTokens <= effectiveLethal) {
                lethalOverride = true;
            }
        }

        // it may be possible in some of the cases that rerollHit + rerollMiss > rerollDiceCount
        if(lethalOverride) {
            rerollMiss = 0;
            rerollHit = 0;
        } else if(rerollMiss === rerollDiceCount) {
            rerollHit = 0;
        } else if(possibleHits <= coverModifier) {
            rerollHit = rerollDiceCount - rerollMiss - effectiveRam;
        } else if(!input.offense.impactX.active && possibleHits <= coverModifier + input.defense.armorX.value) {
            rerollHit = rerollDiceCount - rerollMiss - effectiveRam;
        } else if(input.offense.impactX.active && possibleHits <= coverModifier + input.offense.impactX.value) {
            rerollHit = 0;
        } else if(input.offense.impactX.active && possibleHits <= coverModifier + input.offense.impactX.value + input.defense.armorX.value) {
            rerollHit = possibleHits - (coverModifier + input.offense.impactX.value) - effectiveRam;
        } else { // enough dice to get past the armor and hits to count as damage
            rerollHit = 0;
        }

        // fix possibility of rolling invalid number of dice
        if(rerollHit < 0) {
            rerollHit = 0;
        }
        if(rerollHit + rerollMiss > rerollDiceCount) {
            rerollHit = rerollDiceCount - rerollMiss;
        }

        // return the result
        const indexes = dmm.getRerollIndexes(rerollHit, rerollMiss);
        return indexes.length == 0 ? undefined : indexes;
    }

    shouldHandle(input: T.AttackInput, reason: RS.RerollReason) : boolean {
        return input.defense.armorX.active;
    }
}

export default ArmorXRerollStrategy;
