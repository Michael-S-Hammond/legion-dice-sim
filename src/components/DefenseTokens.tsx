import React from 'react';

import ClearButton from './ClearButton';
import Token from './TokenCounter';
import * as T from '../code/Types'
import * as AS from './AppStateManager';

type DefenseTokensProperties = {
    showSimplifiedView: boolean,
    tokens: T.DefenseTokens,
    eventHandlers: AS.AppStateDefenseEventHandlers,
}

class DefenseTokens extends React.Component<DefenseTokensProperties> {
    constructor(props : DefenseTokensProperties) {
        super(props);
    }

    render() {
        return (
            <div>
                <div className="d-flex justify-content-center my-2">
                    <Token visible={true} value={this.props.tokens.dodge} tokenCssClass='token-counter-dodge' onClick={this.props.eventHandlers.incrementDodgeTokenCount} tooltip='Dodge'></Token>
                    <Token visible={true} value={this.props.tokens.surge} tokenCssClass='token-counter-surge' onClick={this.props.eventHandlers.incrementSurgeTokenCount} tooltip='Surge'></Token>
                    <Token visible={!this.props.showSimplifiedView || this.props.tokens.shield != 0} value={this.props.tokens.shield} tokenCssClass='token-counter-shield' onClick={this.props.eventHandlers.incrementShieldTokenCount} tooltip='Shield'></Token>
                    <Token visible={!this.props.showSimplifiedView || this.props.tokens.observation != 0} value={this.props.tokens.observation} tokenCssClass='token-counter-observation' onClick={this.props.eventHandlers.incrementObservationTokenCount} tooltip='Observation'></Token>
                    <Token visible={true} value={this.props.tokens.suppression} tokenCssClass="token-counter-suppression" onClick={this.props.eventHandlers.incrementSuppressionTokenCount} tooltip='Suppression'></Token>
                    <ClearButton onClick={this.props.eventHandlers.resetTokenCount} tooltip='Clear defense tokens'></ClearButton>
                </div>
            </div>
        );
    }
}

export default DefenseTokens;
