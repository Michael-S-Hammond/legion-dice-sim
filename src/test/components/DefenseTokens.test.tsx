import React from 'react';

import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import * as T from '../../code/Types';
import DefenseTokens from '../../components/DefenseTokens';

import * as EventMocks from '../mocks/EventHandlerMocks';

describe('DefenseTokens', () => {
    it('matches the snapshot', () => {
        const input = T.createDefaultAttackInput();
        const events = EventMocks.createMockAppStateDefenseEventHandlers();

        const defenseAbilities = shallow(<DefenseTokens
            showSimplifiedView={false}
            tokens={input.defense.tokens}
            eventHandlers={events}
            ></DefenseTokens>);

        const snapshot = shallowToJson(defenseAbilities);
        expect(snapshot).toMatchSnapshot();
    });

    it('matches the snapshot for simplified view', () => {
        const input = T.createDefaultAttackInput();
        const events = EventMocks.createMockAppStateDefenseEventHandlers();

        const defenseAbilities = shallow(<DefenseTokens
            showSimplifiedView={true}
            tokens={input.defense.tokens}
            eventHandlers={events}
            ></DefenseTokens>);

        const snapshot = shallowToJson(defenseAbilities);
        expect(snapshot).toMatchSnapshot();
    });
});
