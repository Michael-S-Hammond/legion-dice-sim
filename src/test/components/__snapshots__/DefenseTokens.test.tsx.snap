// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefenseTokens matches the snapshot 1`] = `
<div>
  <div
    className="d-flex justify-content-center my-2"
  >
    <Token
      onClick={[MockFunction]}
      tokenCssClass="token-counter-dodge"
      tooltip="Dodge"
      value={0}
      visible={true}
    />
    <Token
      onClick={[MockFunction]}
      tokenCssClass="token-counter-surge"
      tooltip="Surge"
      value={0}
      visible={true}
    />
    <Token
      onClick={[MockFunction]}
      tokenCssClass="token-counter-shield"
      tooltip="Shield"
      value={0}
      visible={true}
    />
    <Token
      onClick={[MockFunction]}
      tokenCssClass="token-counter-observation"
      tooltip="Observation"
      value={0}
      visible={true}
    />
    <Token
      onClick={[MockFunction]}
      tokenCssClass="token-counter-suppression"
      tooltip="Suppression"
      value={0}
      visible={true}
    />
    <ClearButton
      onClick={[MockFunction]}
      tooltip="Clear defense tokens"
    />
  </div>
</div>
`;

exports[`DefenseTokens matches the snapshot for simplified view 1`] = `
<div>
  <div
    className="d-flex justify-content-center my-2"
  >
    <Token
      onClick={[MockFunction]}
      tokenCssClass="token-counter-dodge"
      tooltip="Dodge"
      value={0}
      visible={true}
    />
    <Token
      onClick={[MockFunction]}
      tokenCssClass="token-counter-surge"
      tooltip="Surge"
      value={0}
      visible={true}
    />
    <Token
      onClick={[MockFunction]}
      tokenCssClass="token-counter-shield"
      tooltip="Shield"
      value={0}
      visible={false}
    />
    <Token
      onClick={[MockFunction]}
      tokenCssClass="token-counter-observation"
      tooltip="Observation"
      value={0}
      visible={false}
    />
    <Token
      onClick={[MockFunction]}
      tokenCssClass="token-counter-suppression"
      tooltip="Suppression"
      value={0}
      visible={true}
    />
    <ClearButton
      onClick={[MockFunction]}
      tooltip="Clear defense tokens"
    />
  </div>
</div>
`;
